File: Benchmark/go/c0_tests.go
package main

import (
	"math"
	"net/http"
	"runtime"
	"sync"
	"time"
)

type C0Response struct {
	Result      float64 `json:"result"`
	DurationMs  float64 `json:"duration_ms"`
	Operations  int     `json:"operations"`
}

// C0a: Vector dot product/reduction (zero allocation)
func c0aVectorDotProduct(state *AppState, w http.ResponseWriter, r *http.Request) {
	size := getQueryParamInt(r, "size", 1000000)
	threads := getQueryParamInt(r, "threads", 1)
	
	// Pre-allocate vectors (not counted as dynamic allocation)
	a := make([]float64, size)
	b := make([]float64, size)
	
	for i := 0; i < size; i++ {
		a[i] = float64(i)
		b[i] = float64(i * 2)
	}
	
	start := time.Now()
	
	var result float64
	
	if threads == 1 {
		// Single-threaded dot product
		for i := 0; i < size; i++ {
			result += a[i] * b[i]
		}
	} else {
		// Multi-threaded
		numCPU := runtime.NumCPU()
		if threads > numCPU {
			threads = numCPU
		}
		
		chunkSize := size / threads
		var wg sync.WaitGroup
		results := make([]float64, threads)
		
		for t := 0; t < threads; t++ {
			wg.Add(1)
			go func(threadID int) {
				defer wg.Done()
				start := threadID * chunkSize
				end := start + chunkSize
				if threadID == threads-1 {
					end = size
				}
				
				var sum float64
				for i := start; i < end; i++ {
					sum += a[i] * b[i]
				}
				results[threadID] = sum
			}(t)
		}
		
		wg.Wait()
		
		for _, partial := range results {
			result += partial
		}
	}
	
	duration := time.Since(start)
	
	response := C0Response{
		Result:     result,
		DurationMs: float64(duration.Nanoseconds()) / 1e6,
		Operations: size,
	}
	
	writeJSONResponse(w, response)
}

// C0b: Vectorizable vs branch-intensive computation
func c0bVectorizableVsBranchy(state *AppState, w http.ResponseWriter, r *http.Request) {
	size := getQueryParamInt(r, "size", 1000000)
	branchy := getQueryParamBool(r, "branchy", false)
	
	data := make([]float64, size)
	for i := 0; i < size; i++ {
		data[i] = float64(i) * 0.1
	}
	
	start := time.Now()
	
	var result float64
	
	if branchy {
		// Branch-intensive version
		for _, x := range data {
			if x < 0.0 {
				result += x * -1.0
			} else if x < 10000.0 {
				result += x*2.0 + 1.0
			} else if x < 50000.0 {
				result += math.Sqrt(x)
			} else {
				result += x / 3.0
			}
		}
	} else {
		// Vectorizable version
		for _, x := range data {
			result += x*2.0 + 1.0
		}
	}
	
	duration := time.Since(start)
	
	response := C0Response{
		Result:     result,
		DurationMs: float64(duration.Nanoseconds()) / 1e6,
		Operations: size,
	}
	
	writeJSONResponse(w, response)
}

// C0c: FFT/Convolution (pre-allocated work area)
func c0cFFTConvolution(state *AppState, w http.ResponseWriter, r *http.Request) {
	size := getQueryParamInt(r, "size", 1024)
	
	// Simple discrete convolution as FFT substitute
	signal := make([]float64, size)
	kernel := []float64{0.25, 0.5, 0.25} // Simple 3-tap filter
	
	for i := 0; i < size; i++ {
		signal[i] = math.Sin(float64(i) * 0.1)
	}
	
	// Pre-allocate output buffer
	output := make([]float64, size)
	
	start := time.Now()
	
	// Convolution operation (zero additional allocation)
	for i := 1; i < size-1; i++ {
		output[i] = signal[i-1]*kernel[0] + signal[i]*kernel[1] + signal[i+1]*kernel[2]
	}
	
	var result float64
	for _, val := range output {
		result += val
	}
	
	duration := time.Since(start)
	
	response := C0Response{
		Result:     result,
		DurationMs: float64(duration.Nanoseconds()) / 1e6,
		Operations: size * len(kernel),
	}
	
	writeJSONResponse(w, response)
}

// C0d: Allocation strategy comparison (pre-allocated vs temporary buffers)
func c0dAllocationStrategy(state *AppState, w http.ResponseWriter, r *http.Request) {
	size := getQueryParamInt(r, "size", 10000)
	usePool := getQueryParamBool(r, "use_pool", true)
	
	start := time.Now()
	
	var result float64
	
	if usePool {
		// Pre-allocated strategy
		buffer := make([]float64, 100)
		
		for i := 0; i < size; i++ {
			// Clear and reuse the same buffer
			for j := range buffer {
				buffer[j] = 0
			}
			
			for j := 0; j < 100; j++ {
				buffer[j] = float64(i + j)
			}
			
			for _, val := range buffer {
				result += val
			}
		}
	} else {
		// Temporary allocation strategy
		for i := 0; i < size; i++ {
			buffer := make([]float64, 100)
			for j := 0; j < 100; j++ {
				buffer[j] = float64(i + j)
			}
			
			for _, val := range buffer {
				result += val
			}
		}
	}
	
	duration := time.Since(start)
	
	response := C0Response{
		Result:     result,
		DurationMs: float64(duration.Nanoseconds()) / 1e6,
		Operations: size * 100,
	}
	
	writeJSONResponse(w, response)
}

File: Benchmark/java/C0Tests.java
public class C0Tests {

    public static class C0Response {
        public double result;
        public double duration_ms;
        public int operations;

        public C0Response(double result, double duration_ms, int operations) {
            this.result = result;
            this.duration_ms = duration_ms;
            this.operations = operations;
        }

        public String toJson() {
            return String.format(
                    "{\"result\":%.12f,\"duration_ms\":%.3f,\"operations\":%d}",
                    result, duration_ms, operations);
        }
    }

    // C0a: Vector dot product/reduction (zero allocation)
    public static C0Response vectorDotProduct(int size, int threads) {
        // Pre-allocate vectors (not counted as dynamic allocation)
        double[] a = new double[size];
        double[] b = new double[size];

        for (int i = 0; i < size; i++) {
            a[i] = i;
            b[i] = i * 2;
        }

        long startTime = System.nanoTime();

        double result;
        if (threads == 1) {
            // Single-threaded dot product
            result = 0.0;
            for (int i = 0; i < size; i++) {
                result += a[i] * b[i];
            }
        } else {
            // Multi-threaded (simplified version)
            int numThreads = Math.min(threads, Runtime.getRuntime().availableProcessors());
            int chunkSize = size / numThreads;
            double[] partialResults = new double[numThreads];
            Thread[] workers = new Thread[numThreads];

            for (int t = 0; t < numThreads; t++) {
                final int threadId = t;
                final int start = threadId * chunkSize;
                final int end = (threadId == numThreads - 1) ? size : (threadId + 1) * chunkSize;

                workers[t] = new Thread(() -> {
                    double sum = 0.0;
                    for (int i = start; i < end; i++) {
                        sum += a[i] * b[i];
                    }
                    partialResults[threadId] = sum;
                });
                workers[t].start();
            }

            // Wait for all threads to complete
            try {
                for (Thread worker : workers) {
                    worker.join();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }

            result = 0.0;
            for (double partial : partialResults) {
                result += partial;
            }
        }

        long endTime = System.nanoTime();
        double duration = (endTime - startTime) / 1_000_000.0; // Convert to milliseconds

        return new C0Response(result, duration, size);
    }

    // C0b: Vectorizable vs branch-intensive computation
    public static C0Response vectorizableVsBranchy(int size, boolean branchy) {
        double[] data = new double[size];
        for (int i = 0; i < size; i++) {
            data[i] = i * 0.1;
        }

        long startTime = System.nanoTime();

        double result = 0.0;
        if (branchy) {
            // Branch-intensive version
            for (double x : data) {
                if (x < 0.0) {
                    result += x * -1.0;
                } else if (x < 10000.0) {
                    result += x * 2.0 + 1.0;
                } else if (x < 50000.0) {
                    result += Math.sqrt(x);
                } else {
                    result += x / 3.0;
                }
            }
        } else {
            // Vectorizable version
            for (double x : data) {
                result += x * 2.0 + 1.0;
            }
        }

        long endTime = System.nanoTime();
        double duration = (endTime - startTime) / 1_000_000.0;

        return new C0Response(result, duration, size);
    }

    // C0c: FFT/Convolution (pre-allocated work area)
    public static C0Response fftConvolution(int size) {
        // Simple discrete convolution as FFT substitute
        double[] signal = new double[size];
        double[] kernel = { 0.25, 0.5, 0.25 }; // Simple 3-tap filter

        for (int i = 0; i < size; i++) {
            signal[i] = Math.sin(i * 0.1);
        }

        // Pre-allocate output buffer
        double[] output = new double[size];

        long startTime = System.nanoTime();

        // Convolution operation (zero additional allocation)
        for (int i = 1; i < size - 1; i++) {
            output[i] = signal[i - 1] * kernel[0] + signal[i] * kernel[1] + signal[i + 1] * kernel[2];
        }

        double result = 0.0;
        for (double val : output) {
            result += val;
        }

        long endTime = System.nanoTime();
        double duration = (endTime - startTime) / 1_000_000.0;

        return new C0Response(result, duration, size * kernel.length);
    }

    // C0d: Allocation strategy comparison (pre-allocated vs temporary buffers)
    public static C0Response allocationStrategy(int size, boolean usePool) {
        long startTime = System.nanoTime();

        double result = 0.0;

        if (usePool) {
            // Pre-allocated strategy
            double[] buffer = new double[100];

            for (int i = 0; i < size; i++) {
                // Clear and reuse the same buffer
                for (int j = 0; j < 100; j++) {
                    buffer[j] = i + j;
                }

                for (double val : buffer) {
                    result += val;
                }
            }
        } else {
            // Temporary allocation strategy
            for (int i = 0; i < size; i++) {
                double[] buffer = new double[100];
                for (int j = 0; j < 100; j++) {
                    buffer[j] = i + j;
                }

                for (double val : buffer) {
                    result += val;
                }
            }
        }

        long endTime = System.nanoTime();
        double duration = (endTime - startTime) / 1_000_000.0;

        return new C0Response(result, duration, size * 100);
    }
}
